File: Packages/Lexicon/Sources/Lexicon/Lexicon.swift
/// Frequency lexicon for ranking and prediction.
///
/// Distinct from `LemmaCore.BinaryLemmatizer` (which answers "is this a
/// valid Icelandic form, and what's its lemma/morphology") — a `Lexicon`
/// answers "how common is this word / this word pair", which is what
/// autocorrect ranking and next-word prediction need.
public protocol Lexicon: Sendable {
    /// Unigram frequency of a (lowercased) word; nil if unknown.
    func frequency(of word: String) -> UInt32?
    /// Frequency of the bigram "first second"; nil if unseen.
    func bigramFrequency(_ first: String, _ second: String) -> UInt32?
    /// Up to `limit` known words starting with `prefix`, descending frequency.
    func completions(of prefix: String, limit: Int) -> [(word: String, frequency: UInt32)]
    /// Up to `limit` words that follow `word` in the bigram table, descending bigram frequency.
    func continuations(of word: String, limit: Int) -> [(word: String, frequency: UInt32)]
    /// Sum of all unigram frequencies (for probability normalization).
    var totalUnigramTokens: UInt64 { get }
}

public extension Lexicon {
    /// Default no-op implementation so conformers that predate this API
    /// (and packages built against an older `Lexicon` snapshot) keep
    /// compiling without change. `FrequencyLexicon` overrides this with a
    /// real bigram-range scan; other conformers may opt in later.
    func continuations(of word: String, limit: Int) -> [(word: String, frequency: UInt32)] {
        []
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/lexicon/Lexicon.kt
package `is`.solberg.lyklabord.engine.lexicon

/**
 * Frequency lexicon for ranking and prediction.
 *
 * Distinct from `BinaryLemmatizer` (which answers "is this a valid Icelandic
 * form, and what's its lemma/morphology") — a `Lexicon` answers "how common is
 * this word / this word pair", which is what autocorrect ranking and next-word
 * prediction need.
 *
 * 1:1 port of `Packages/Lexicon/Sources/Lexicon/Lexicon.swift`.
 */
interface Lexicon {
    /** Unigram frequency of a (lowercased) word; null if unknown. */
    fun frequency(of: String): UInt?

    /** Frequency of the bigram "first second"; null if unseen. */
    fun bigramFrequency(first: String, second: String): UInt?

    /** Up to [limit] known words starting with [prefix], descending frequency. */
    fun completions(of: String, limit: Int): List<LexEntry>

    /** Up to [limit] words that follow [word] in the bigram table, descending bigram frequency. */
    fun continuations(of: String, limit: Int): List<LexEntry> = emptyList()

    /** Sum of all unigram frequencies (for probability normalization). */
    val totalUnigramTokens: ULong
}

/** A `(word, frequency)` pair, mirroring the Swift tuple `(word:frequency:)`. */
data class LexEntry(val word: String, val frequency: UInt)
